### PR TITLE
C++: Add ECMAScript `std::regex` parser

### PR DESCRIPTION
**Pull request by jketema**
Commit-by-commit review recommended. This is not hooked into anything yet, as this is getting quite big already, so I didn't want to add the flow config here, or other regex alternatives that can be used with `std::regex`.

Approach:
1. Import the Ruby parser, make QL and tests compile, fix position issue with raw strings
2. Remove unsupported features
3. Add features that are not supported by the Ruby parser
4. Add missing test cases
5. Some after-the-fact fixes.

### CODE REVIEW

File: cpp/ql/test/library-tests/regex/regexp.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+std::regex r_alt("foo|bar");
+
+// Character classes
+std::regex r_cc1("[abc]");
+std::regex r_cc2("[a-fA-F0-9_]");
+std::regex r_cc3("[\\w]+");
+std::regex r_cc4("\\[\\][123]");
+std::regex r_cc5("[^A-Z]");
+std::regex r_cc6("[]]");   // MRI gives a warning, but accepts this as matching ']'
+std::regex r_cc7("[^]]"); // MRI gives a warning, but accepts this as matching anything except ']'
```
Comment (geoffw0) [inline cpp/ql/test/library-tests/regex/regexp.cpp]: What is MRI?

File: cpp/ql/test/library-tests/regex/regexp.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+std::basic_regex<char8_t> r_loc_u8(u8"a\\nc");
+std::basic_regex<char16_t> r_loc_u(u"a\\nc");
+std::basic_regex<char32_t> r_loc_U(U"a\\nc");
+std::regex r_loc_R(R"(a\nc)");
+std::basic_regex<wchar_t> r_loc_LR(LR"(a\nc)");
+std::basic_regex<char8_t> r_loc_u8R(u8R"(a\nc)");
+std::basic_regex<char16_t> r_loc_uR(uR"(a\nc)");
+std::basic_regex<char32_t> r_loc_UR(UR"(a\nc)");
+std::regex r_loc_Rx(R"x(a\nc)x");
+std::regex r_loc_Rfoo(R"foo(a\nc)foo");
```
Comment (geoffw0) [inline cpp/ql/test/library-tests/regex/regexp.cpp:125]: Should we have test cases for invalid regexs to ensure nothing too strange happens there - and for that matter string literals that are not remotely regexs, to ensure they are not misidentified as such.

File: cpp/ql/lib/semmle/code/cpp/exprs/Literal.qll [hunk 3]
```diff
@@ -133,6 +133,19 @@ class StringLiteral extends TextLiteral {
...
 }
 
+/**
+ * A C++ raw string literal. For example:
+ * ```
+ * const char *s1 = R"(abcdef)";
+ * const wchar_t *s2 = LR"x(123456)x";
+ * ```
+ */
+class RawStringLiteral extends StringLiteral {
```
Comment (geoffw0) [inline cpp/ql/lib/semmle/code/cpp/exprs/Literal.qll:143]: It might be nice to have a test for `RawStringLiteral` itself, even if it's just a very basic test as part of the regex tests e.g.
```
from RawStringLiteral rsl
select rsl
```

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll [hunk 4]
```diff
@@ -0,0 +1,1247 @@
...
+     */
+    private int regexpContentOffset() {
+      exists(string vt | vt = re.getValueText() |
+        // Find the '(' that opens the raw content.
+        re instanceof RawStringLiteral and
+        result = 1 + min(int i | vt.charAt(i) = "(")
+        or
+        // Find the opening '"'.
+        not re instanceof RawStringLiteral and
+        result = 1 + min(int i | vt.charAt(i) = "\"")
```
Comment (geoffw0) [inline cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: It might be nice to make this a method of `StringLiteral`, something like `getContentOffset`, and overload it in `RawStringLiteral`?

**Review by geoffw0 — COMMENTED** [review]
Review part 1.  Thank you for breaking this PR into logical commits, with commit comments explaining the changes - this was extremely helpful for making sense of the changes you've made from the Ruby implementation I'm already somewhat familiar with.

I've also created a small PR https://github.com/jketema/codeql/pull/33 onto this PR, addressing an edge case with string locations that I investigated, and figured I might as well trivially fix and share my code changes for.

File: cpp/ql/test/library-tests/regex/regexp.cpp (hunk 1 above)
Comment (jketema) [reply cpp/ql/test/library-tests/regex/regexp.cpp]: https://en.wikipedia.org/wiki/Ruby_MRI

This is a left-over from Ruby tests that I copied. We should probably just remove the comments.

File: cpp/ql/test/library-tests/regex/regexp.cpp (hunk 2 above)
Comment (jketema) [reply cpp/ql/test/library-tests/regex/regexp.cpp:125]: Is there anything specific you're thinking about?

File: cpp/ql/lib/semmle/code/cpp/exprs/Literal.qll (hunk 3 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/exprs/Literal.qll:143]: Makes sense. Will add.

**Comment by jketema** [comment]
> I've also created a small PR [jketema#33](https://github.com/jketema/codeql/pull/33) onto this PR, addressing an edge case with string locations that I investigated, and figured I might as well trivially fix and share my code changes for.

Commented on that here: https://github.com/jketema/codeql/pull/33/files#r3871704321

File: cpp/ql/test/library-tests/regex/regexp.cpp (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/test/library-tests/regex/regexp.cpp:125]: Nothing in particular, just (1) sometimes people write invalid regexs, and sometimes libraries even accept them so they don't get spotted; (2) real world databases will contain a lot of string literals that aren't regexs, so it's important we don't do anything silly (or computationally expensive for that matter) with them.  I can't remember how this worked in Ruby / Python / Swift.

Comment (jketema) [reply cpp/ql/test/library-tests/regex/regexp.cpp:125]: > I can't remember how this worked in Ruby / Python / Swift.

I think the idea is that we only look at strings that actually flow into the correct `std::regex` function arguments. That is clearly missing at this point.

File: cpp/ql/test/library-tests/regex/regexp.cpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+namespace std {
+  template <class CharT>
+  class basic_regex {
+  public:
+    basic_regex(const CharT *s) {}
+    basic_regex(const CharT *s, int flags) {}
+    basic_regex &assign(const CharT *s) { return *this; }
+  };
+  typedef basic_regex<char> regex;
```
Comment (geoffw0) [inline cpp/ql/test/library-tests/regex/regexp.cpp:11]: Do we want to test `std::wregex` as well?

And initialization constants such as `std::regex_constants::icase`? (which I think we can support fairly easily, when we're ready to add that)

And non-ECMA regexs? (which I guess, realistically, we plan to either ignore for the time being???)

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll [hunk 6]
```diff
@@ -1063,47 +1061,9 @@ private module Impl implements RegexTreeViewSig {
...
-  /**
-   * A zero-width lookbehind assertion.
-   *
-   * Examples:
-   *
-   * ```
-   * (?<=\.)
-   * (?<!\\)
-   * ```
-   */
```
Comment (geoffw0) [inline cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: I'm a bit surprised lookbehind isn't supported in ECMAScript regexs.

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll [hunk 7]
```diff
@@ -1112,12 +1112,6 @@ private module Impl implements RegexTreeViewSig {
      * `"digit"`.
      */
     string getName() { result = re.getCharacterPropertyName(start, end) }
-
-    /**
-     * Holds if the property is inverted. For example, it holds for `[[:^digit:]]`,
-     * which matches non-digits.
```
Comment (geoffw0) [inline cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: Again, I'm a bit surprised this isn't supported.

File: cpp/ql/test/library-tests/regex/regexp.cpp [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+std::regex r_posix1("[[:alpha:]][[:digit:]]");
+
+// A single character class containing two POSIX bracket expressions
+std::regex r_posix2("[[:alpha:][:digit:]]");
+
+// A single character class containing two ranges and one POSIX bracket expression
+std::regex r_posix3("[A-F[:digit:]a-f]");
+
+// POSIX collating symbols
+std::regex r_posix_coll1("[[.a.]]");
```
Comment (geoffw0) [inline cpp/ql/test/library-tests/regex/regexp.cpp:81]: If I'm understanding correctly (I'm not confident I am) this is a somewhat strange example - since the collating symbol is intended to join _several_ characters as though they are one?

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll [hunk 9]
```diff
@@ -0,0 +1,1247 @@
...
+    RegExpEscape() { re.escapedCharacter(start, end) }
+
+    /**
+     * Gets the name of the escaped; for example, `w` for `\w`.
+     * TODO: Handle named escapes.
+     */
+    override string getValue() {
+      not this.isUnicode() and
+      not this.isHex() and
+      not this.isControl() and
```
Comment (geoffw0) [inline cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: I believe these are redundant, as they're checked inside `isIdentityEscape`.
```suggestion
```

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll [hunk 10]
```diff
@@ -0,0 +1,1247 @@
...
+    }
+
+    override string getAPrimaryQlClass() { result = "RegExpCharacterRange" }
+
+    override predicate isNullable() { none() }
+  }
+
+  /**
+   * A normal character in a regular expression, that is, a character
+   * without special meaning. This includes escaped characters.
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: Rename to `RegExpNormalChars`???

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll [hunk 11]
```diff
@@ -0,0 +1,1247 @@
...
+  /**
+   * Holds if `root` has the `i` flag for case-insensitive matching.
+   */
+  predicate isIgnoreCase(RegExpTerm root) {
+    root.isRootTerm() and
+    root.getLiteral().isIgnoreCase()
+  }
+
+  /**
+   * Holds if `root` has the `s` flag for multi-line matching.
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: Yep, this appears to be a stale comment.

**Review by geoffw0 — COMMENTED** [review]
Review part 2.  I've now reviewed all of the code and tests.

I'm very happy with what I see, you've clearly put considerably effort into getting this right ... though experience suggests we should be prepared to find bugs and missing features when we start to implement queries on top of the library, as doing so will test it more deeply.

I'm going to do a few quick tests locally, but I'm leaning towards getting this merged quickly so we can build on it.

:+1:

File: cpp/ql/test/library-tests/regex/regexp.cpp (hunk 5 above)
Comment (jketema) [reply cpp/ql/test/library-tests/regex/regexp.cpp:11]: `icase` requires dataflow.

non-ECMA regexes are on my todo list but only after I've rigged up whatever else I need to get to a sensible Redos query. These additional regex formats seem relatively easy to add, but again some dataflow will be required to determine which regex format is being used.

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll (hunk 6 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: Current ECMAScript does support it as far as I recall when looking into this, but it is not supported in the specific version that the C++ standard took as its starting point.

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll (hunk 7 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: I think this is pretty Ruby specific?

File: cpp/ql/test/library-tests/regex/regexp.cpp (hunk 8 above)
Comment (jketema) [reply cpp/ql/test/library-tests/regex/regexp.cpp:81]: That might be. I'll have a look and update the test if appropriate.

File: cpp/ql/test/library-tests/regex/regexp.cpp (hunk 5 above)
Comment (jketema) [reply cpp/ql/test/library-tests/regex/regexp.cpp:11]: > Do we want to test `std::wregex` as well?

I'm not sure that's adding much at this stage, but let me add the needed prototypes and at least one test.

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll (hunk 9 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: You might be right. I'll check.

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll (hunk 11 above)
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: I'm not actually sure about that. The predicate name seems very specific to one specific regex version parsed in one of the languages where we have a similar regex library. I'll double check though.

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll (hunk 6 above)
Comment (geoffw0) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: Ah, that explains the conflicting information I was seeing on this.

Thanks.

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll (hunk 7 above)
Comment (geoffw0) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: On https://en.cppreference.com/cpp/regex/ecmascript it has an example `[^[:digit:]]`, but that's slightly different, and looking at it now I think it's just a regular use of `[^...]`.

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: Indeed. `[^[:digit:]]` should be covered by the regular `[^...]` case. I'll add a test case of that if there isn't one already.

**Comment by geoffw0** [comment]
> I'm going to do a few quick tests locally

I've now done my local testing.  I noticed locations of regex terms where the regex includes one or more escape sequences (`\\`) are a little bit off.  It doesn't need fixing now, but it's worth noting I think.  Other than that I'm happy.

---

Nothing else to review, I'm happy to approve this when everything above has been answered.  I don't think anything I've asked for is critical at this stage (as no queries use the library yet), so feel free to write up a TODO list issue instead of fixing every detail now.  It will be good to have some code in `main`.  🚀

**Comment by jketema** [comment]
> I noticed locations of regex terms where the regex includes one or more escape sequences (`\\`) are a little bit off.

I saw that. I tried fixing that, but that turned out to be non-trivial. I'll add a note in the `qll` file on this.

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll [hunk 12]
```diff
@@ -0,0 +1,1247 @@
...
+    }
+
+    override RegExpTerm getChild(int i) {
+      i = 0 and
+      result.getRegExp() = re and
+      result.getStart() = start and
+      result.getEnd() = part_end
+    }
+
+    /** Hodls if this term may match an unlimited number of times. */
```
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: This was fixed in b15082f9e27f51087c3431191e86ab5dcc51e49c (already reviewed).

File: cpp/ql/test/library-tests/regex/regexp.cpp (hunk 1 above)
Comment (jketema) [reply cpp/ql/test/library-tests/regex/regexp.cpp]: I've updated the comments to `No a valid regex` in d9ad582a087ffc8b21cf1e0e945f8d49c939b2d8. Clang's regex library rejects both at runtime with:
```
The parser did not consume the entire regular expression.
```
I did not test gcc, but would expect similar behavior.

I'm assuming that correct regexes need to be parsed correctly, but that for incorrect ones it doesn't really matter, as those should generally not occur because they lead to runtime exceptions in actual code.

File: cpp/ql/test/library-tests/regex/regexp.cpp (hunk 2 above)
Comment (jketema) [reply cpp/ql/test/library-tests/regex/regexp.cpp:125]: I left this as-is under the assumption I stated above: I'm assuming that correct regexes need to be parsed correctly, but that for incorrect ones it doesn't really matter, as those should generally not occur because they lead to runtime exceptions in actual code.

> (1) sometimes people write invalid regexs, and sometimes libraries even accept them so they don't get spotted;

This is fair. I think that if we find cases where the library accepts them, then we definitely need to update the parser to handle them correctly.

> (2) real world databases will contain a lot of string literals that aren't regexs

Agreed. For some for context, note that `RegExp` is abstract. In the tests I currently do:
```
class RegExpTest extends RegExp {
  RegExpTest() { any() }
}
```
Follow-up work should extend `RegExp` in the library in such a way that the charpred restricts this to only strings that flow into `std:regex` and friends.

File: cpp/ql/lib/semmle/code/cpp/exprs/Literal.qll (hunk 3 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/exprs/Literal.qll:143]: Fixed in fed331e2666

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll (hunk 11 above)
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: Fixed in e6359c596dd. I observe that this is coming from the `RegexTreeViewSig` signature, which also has a similar (incorrect) comment.

Note that dot-all actually doesn't make a lot of sense in the case of C++ (not a grammar variation option, see https://cppreference.com/cpp/regex/basic_regex) and that we will need to add machinery for multi-line regexes (that is a grammar variation). Note that the latter will require dataflow.

File: cpp/ql/lib/semmle/code/cpp/regex/internal/ParseRegExp.qll [hunk 13]
```diff
@@ -0,0 +1,871 @@
...
+private import semmle.code.cpp.exprs.Literal
+
+/**
+ * A C++ string literal containing a regular expression term, that is, either
+ * a regular expression literal, or a string literal used in a context where
+ * it is parsed as a regular expression.
+ */
+abstract class RegExp extends StringLiteral {
+  /**
+   * Holds if this `RegExp` has the `s` flag for multi-line matching.
```
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/regex/internal/ParseRegExp.qll]: Fixed in e6359c596dd. See https://github.com/github/codeql/pull/22300#discussion_r3881433990.

File: cpp/ql/test/library-tests/regex/regexp.cpp (hunk 5 above)
Comment (jketema) [reply cpp/ql/test/library-tests/regex/regexp.cpp:11]: `std:wregex` added in bd1fc196b07 and updated the one test case that we effectively already had for this. This unfortunately shifts quite a lot around in the test. I verified that this is correct by initially putting `wregex` on the same line as `regex` and observing that the update from `std::basic_regex<wchar_t>` to `std:wregex` only updated locations related to that line.

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll (hunk 7 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: Test added in 957d1fcc455. Parsing looks correct:
```
#   69| [RegExpCharacterClass] [^[:digit:]]
#-----| 0 -> [RegExpNamedCharacterProperty] [:digit:]
```

File: cpp/ql/test/library-tests/regex/regexp.cpp (hunk 8 above)
Comment (jketema) [reply cpp/ql/test/library-tests/regex/regexp.cpp:81]: Clang's regex parser accepts `[[.a.]]`; I agree it is somewhat strange. In 7732801d6d0 I've added `[[.ch.]]` as a slightly less artificial example.

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll (hunk 9 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: Fixed in 18074d07fbe

**Comment by jketema** [comment]
> I noticed locations of regex terms where the regex includes one or more escape sequences (`\\`) are a little bit off.

Added a comment in 40a87469450.

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll (hunk 10 above)
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: Fixed in 0ab94543fb. Note that this causes test changes.

**Comment by jketema** [comment]
> I've also created a small PR [jketema#33](https://github.com/jketema/codeql/pull/33) onto this PR, addressing an edge case with string locations that I investigated, and figured I might as well trivially fix and share my code changes for.

I've cherry-picked your commits and resolved expected test result conflicts in the process. They occur here as https://github.com/github/codeql/pull/22300/commits/9e52f2836be976c88a5576ad6aec9683d7b33b71 and https://github.com/github/codeql/pull/22300/commits/8e169d351b11ab5534020b26755eb4afc704320e.

File: cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll [hunk 14]
```diff
@@ -0,0 +1,1254 @@
...
+        re instanceof RawStringLiteral and
+        result = 1 + min(int i | vt.charAt(i) = "(")
+        or
+        // Find the opening '"'.
+        not re instanceof RawStringLiteral and
+        result = 1 + min(int i | vt.charAt(i) = "\"")
+        or
+        // No opening '"'.
+        not re instanceof RawStringLiteral and
+        not exists(int i | vt.charAt(i) = "\"") and
```
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/regex/RegexTreeView.qll]: Fixed in https://github.com/github/codeql/pull/22300/commits/52f171078dd5aa39384dc88b87045ec8fc62c672

**Comment by jketema** [comment]
State as of this writing: still need to have a look at the following from https://github.com/jketema/codeql/pull/33, otherwise all comments have been addressed.

> I think we should also consider moving most of the logic from `regexpContentOffset()` into a method `StringLiteral.getContentOffset()`, overridden in `RawStringLiteral` for the additional logic of that situation.

Expanded on the fragility of the locations here: https://github.com/github/codeql/pull/22300/commits/50ed141605c8c428c53b64fdcc9f4fd144e69da3

**Comment by jketema** [comment]
> I think we should also consider moving most of the logic from `regexpContentOffset()` into a method `StringLiteral.getContentOffset()`, overridden in `RawStringLiteral` for the additional logic of that situation.

I'm going to leave this in the regex code for now. The code is pretty fragile as far as macro uses are concerned, so I prefer to leave it out of the core of the library.

**Comment by jketema** [comment]
For completeness: I believe I addressed all comments.
